### PR TITLE
EDM-613: Remove redundant config discriminator

### PR DESCRIPTION
**Pull request by fzdarsky**
Our `device.config` API currently uses discriminators to determine the config provider type, e.g.

```
- name: a-git-config
  configType: GitConfigProviderSpec
  gitRef:
    repository: default-repo
    [...]
```

The `configType` is a discriminator that leaks an internal schema name (`GitConfigProviderSpec`) and is both clunky from user perspective and redundant, as the schema can be inferred from the `gitRef` property as we do with other schemas (e.g. `ApplicationSpec`):

```
- name: a-git-config
  gitRef:
    repository: default-repo
    [...]
```

### CODE REVIEW

**Comment by celdrake** [comment]
@fzdarsky The UI currently relies on the field for managing the configurations.
I will make the necessary adaptations to go with this PR tomorrow, as it's been used in a few places.

File: api/v1alpha1/util.go [hunk 1]
```diff
@@ -40,6 +49,28 @@ func (t HookAction) Type() (HookActionType, error) {
 	return "", fmt.Errorf("unable to determine action type: %+v", data)
 }
 
+// Type returns the type of the application provider.
```
Comment (hexfusion) [inline api/v1alpha1/util.go:52]: nit: `s/application/config` :)

Comment (fzdarsky) [reply api/v1alpha1/util.go]: Argh. Good catch. Fixed.

**Comment by fzdarsky** [comment]
> @fzdarsky The UI currently relies on the field for managing the configurations. I will make the necessary adaptations to go with this PR tomorrow, as it's been used in a few places.

Thanks @celdrake. IDK your backlog for v0.3. It'd be nice to get this in, but no huge issue if we have to bump to next. Please decide at your own discretion.

File: internal/tasks/templateversion_populate.go [hunk 2]
```diff
@@ -110,23 +110,23 @@ func (t *TemplateVersionPopulateLogic) getFleetAndTemplateVersion(ctx context.Co
...
-	disc, err := configItem.Discriminator()
+func (t *TemplateVersionPopulateLogic) handleConfigItem(ctx context.Context, configItem *api.ConfigProviderSpec) error {
+	configType, err := configItem.Type()
 	if err != nil {
-		return fmt.Errorf("failed getting discriminator: %w", err)
+		return fmt.Errorf("failed getting config type: %w", err)
 	}
 
-	switch disc {
-	case string(api.TemplateDiscriminatorGitConfig):
```
Comment (celdrake) [inline internal/tasks/templateversion_populate.go:120]: Nit: The types `TemplateDiscriminator`, `TemplateDiscriminatorGitConfig` etc can be removed.
They are defined in the API and in the openapi spec.

**Review by celdrake — COMMENTED** [review]
@fzdarsky The PR in the UI side is ready, I think we can include this story in v0.3

File: internal/tasks/templateversion_populate.go (hunk 2 above)
Comment (fzdarsky) [reply internal/tasks/templateversion_populate.go:120]: Removed, thank you.

File: api/v1alpha1/validation.go [hunk 3]
```diff
@@ -25,12 +25,7 @@ func (r Device) Validate() []error {
 		}
 		if r.Spec.Config != nil {
 			for _, config := range *r.Spec.Config {
-				value, err := config.ValueByDiscriminator()
-				if err != nil {
-					allErrs = append(allErrs, fmt.Errorf("invalid configType: %s", err))
-				} else {
-					allErrs = append(allErrs, value.(Validator).Validate()...)
-				}
+				allErrs = append(allErrs, config.Validate()...)
```
Comment (hexfusion) [inline api/v1alpha1/validation.go:28]: :+1:

File: examples/fleet-b.yaml [hunk 4]
```diff
@@ -15,20 +15,17 @@ spec:
         image: quay.io/redhat/rhde:9.2
       config:
         - name: base
-          configType: GitConfigProviderSpec
```
Comment (hexfusion) [inline examples/fleet-b.yaml:18]: :heart:

**Review by hexfusion — APPROVED** [review]
nice

**Pull request merged**